### PR TITLE
More flexible configuration

### PR DESCRIPTION
**Pull request by MikiDi**
Many thanks for this initiative. I was surprised to find this is the only initiative (I could find) that came to life after the [Mapillary events](https://lists.openstreetmap.org/pipermail/talk/2020-June/084985.html). Have been pondering on something like this for quite a while myself. Very excited about the ambition for distributed hosting as well! (has any discussion/thinking on this subject been done already? If so, I'd be happy to join in ...)

Since I'm not very familiar with php-based environments, I decided to go ahead and dockerize the example apps in order to make deployment easier for someone that doesn't have all the necessary components installed on their machine. To achieve that, I needed to make some minor changes to the way the server can be configured:

- make setting of env variables through an `.env`-file optional
- make database hostname and password configurable (the [most common postgis docker-images](https://hub.docker.com/r/kartoza/postgis/) require a password :worried:  ) 

I'll be sending in a (draft) PR on the frontend examples soon.

PS: @nickw1 Cool to see how open communities can be interconnected (Recognized your name from the [AR.js PR](https://github.com/AR-js-org/AR.js/pull/116) :blush:  )

### CODE REVIEW

**Comment by nickw1** [comment]
Hi @MikiDi many thanks for your PR! Yes, I recognise you as the main contributor of the ES6 port in AR.js, glad you are interested in this project! To give a bit of history: I've been working for some time (more seriously in the past two years) on OpenTrailView (https://gitlab.com/nickw1/opentrailview), a FOSS panorama navigation app for off-road trails. Then last summer, after the Mapillary thing happened I got talking to my collaborator, @mrAceT and we started the OpenWanderer project. Only preliminary discussions so far on the distributed aspect, but that's what we'd like to go for. I too am surprised there are (apparently) no other similar projects at the moment.

I can see one issue with your change in `OpenWanderer.php`; you use a variable `$envDir` which isn't defined. I'm happy to fix that myself; as soon as that is done I will merge.

Docker containers would be most welcome. Docker is one of the (many) things I've been meaning to get into for a while, but have not had the time just yet. However it would obviously be useful for this project; for one thing making the server itself a Docker container would be useful I think.

Another thing I may do if I have the time is implement a node.js version of the server, so that people who are happy working with node.js but not with PHP (which these days will be quite a few, I'd imagine) can deploy easily. Do you have any feelings on that?

Basically the more people who can host an instance, the better.just need to get more people interested! Thanks for offering to be part of the hosting when the time comes for that.

**Comment by MikiDi** [comment]
Thanks for the quick response! I'm finalizing the docker-setup of the example app over here. 

> Only preliminary discussions so far on the distributed aspect, but that's what we'd like to go for.

> Thanks for offering to be part of the hosting when the time comes for that.

To clarify: This question was meant primarily from a technical point of view, eg. "What has been done so far, so I can consider that before helping (designing & developing) to try and make this a reality?". That is íf you're looking for contributors ofcourse ... When the time comes I'd participate in hosting as well :wink:

> Another thing I may do if I have the time is implement a node.js version of the server, so that people who are happy working with node.js but not with PHP (which these days will be quite a few, I'd imagine) can deploy easily. Do you have any feelings on that?

Yeah ... It's not so much for the deployment part that it's more difficult (once in Docker at least :whale: :wink:  ), but for attracting contributors one might indeed be better off with Node.js these days. That would apply for me as well. Before implementing a new server however, I was going to propose converting the API documentation to something more structured such as [OpenAPI](https://spec.openapis.org/oas/v3.1.0) (Swagger). Having something like that reads even more easily and has the additional benefit that it can be loaded into an interactive request-generator such as Swagger to experiment (It's all about lowering the bar for contributors to experiment, right?). This all to say that defining the API is arguably the hardest part. Once you have something that might work well, it's "just" a matter of implementing (in a language/framework of choice).

**Comment by mrAceT** [comment]
Welcome @MikiDi !

I think one of the reasons that something like this hasn't surfaced is that to implement it is quite a thing to do.. I was very happy to find "a listening ear" (literately translated Dutch proverb) at @nickw1 .

To quote Thomas Edison: Genius is 1% inspiration and 99% perspiration. We both mixed our inspirations into this concept, but I must admit that @nickw1 is doing most of the perspiration at the moment ;)

I got talking with Nick because of his project OpenTrailView and I am working on a (for now private) project I've called GeoArchive. We've opensourced the basic concept and we've been talking about finding a way to give it the potential to give the concept a robust future.. also hoping we'dd attract other developers ;)

**Comment by nickw1** [comment]
> Thanks for the quick response! I'm finalizing the docker-setup of the example app over here.
> 
> > Only preliminary discussions so far on the distributed aspect, but that's what we'd like to go for.
> 
> > Thanks for offering to be part of the hosting when the time comes for that.
> 
> To clarify: This question was meant primarily from a technical point of view, eg. "What has been done so far, so I can consider that before helping (designing & developing) to try and make this a reality?". That is íf you're looking for contributors ofcourse ... When the time comes I'd participate in hosting as well wink

Ah, ok - sorry for misunderstanding. I have to admit we haven't got into the details of the technical side of the distributed hosting yet... but yes, we would most definitely welcome contributors so feel free to help on this should you wish to ;-)

> > Another thing I may do if I have the time is implement a node.js version of the server, so that people who are happy working with node.js but not with PHP (which these days will be quite a few, I'd imagine) can deploy easily. Do you have any feelings on that?
> 
> Yeah ... It's not so much for the deployment part that it's more difficult (once in Docker at least whale wink ), 

Was thinking that, hence your Docker contributions being so welcome ;-)

>but for attracting contributors one might indeed be better off with Node.js these days. That would apply for me as well. Before implementing a new server however, I was going to propose converting the API documentation to something more structured such as [OpenAPI](https://spec.openapis.org/oas/v3.1.0) (Swagger). Having something like that reads even more easily and has the additional benefit that it can be loaded into an interactive request-generator such as Swagger to experiment (It's all about lowering the bar for contributors to experiment, right?). This all to say that defining the API is arguably the hardest part. Once you have something that might work well, it's "just" a matter of implementing (in a language/framework of choice).

Sounds a good plan. @mrAceT what are your feelings about a Node.js server (I know you've expressed a preference for PHP in the past, but we might get more developers with Node as @MikiDi says). Node/Express isn't too hard to pick up if you know server-side coding and client-side JS, however we can always develop PHP and Node-based servers in parallel if need be.

**Comment by MikiDi** [comment]
> Welcome @MikiDi !

Hi @mrAceT , :belgium: here. I handle literally translated probverbs pretty well :wink: 

> I got talking with Nick because of his project OpenTrailView and I am working on a (for now private) project I've called GeoArchive. We've opensourced the basic concept and we've been talking about finding a way to give it the potential to give the concept a robust future.. also hoping we'dd attract other developers ;)

Figured out most of that because of your [posts on trekview forum](https://campfire.trekview.org/t/an-open-source-street-level-mapping-platform-viable/238) :relaxed: 

It'd be nice to have a place to collect these posts/articles talking about the project in some kind of a "links"-readme (the link to the original OSM mailing list conversation, prior art in open sotware, ...). Useful as background for newcomers ... Would the https://github.com/openwanderer/openwanderer repo be a good place?

**Comment by nickw1** [comment]
@MikiDi yes, I think the [openwanderer/openwanderer](https://github.com/openwanderer/openwanderer) repo would be a good place to put these. There is also a plan to setup a wiki too at some point... @mrAceT is going to do that when he has a moment.
